src/Meshes/mesh.vertexData.ts

```diff
@@ -568,58 +577,73 @@ export class VertexData {
         if (isAsync) { yield; }
-        this.matricesWeights = VertexData._mergeElement(this.matricesWeights, others.map((other) => other.matricesWeights));
+        this.matricesWeights = VertexData._mergeElement(VertexBuffer.MatricesWeightsKind, this.matricesWeights, transform, vertexDatas.map((other) => [other[0].matricesWeights, other[1]]));
         if (isAsync) { yield; }
-        this.matricesIndicesExtra = VertexData._mergeElement(this.matricesIndicesExtra, others.map((other) => other.matricesIndicesExtra));
+        this.matricesIndicesExtra = VertexData._mergeElement(VertexBuffer.MatricesIndicesExtraKind, this.matricesIndicesExtra, transform, vertexDatas.map((other) => [other[0].matricesIndicesExtra, other[1]]));
         if (isAsync) { yield; }
-        this.matricesWeightsExtra = VertexData._mergeElement(this.matricesWeightsExtra, others.map((other) => other.matricesWeightsExtra));
+        this.matricesWeightsExtra = VertexData._mergeElement(VertexBuffer.MatricesWeightsExtraKind, this.matricesWeightsExtra, transform, vertexDatas.map((other) => [other[0].matricesWeightsExtra, other[1]]));
 
         return this;
     }
 
-    private static _mergeElement(source: Nullable<FloatArray>, others: readonly Nullable<FloatArray>[]): Nullable<FloatArray> {
-        const nonNullOthers = others.filter((other): other is FloatArray => other !== null && other !== undefined);
+    private static _mergeElement(kind: string, source: Nullable<FloatArray>, transform: Matrix | undefined, others: readonly (readonly [element: Nullable<FloatArray>, transform?: Matrix])[]): Nullable<FloatArray> {
+        const nonNullOthers = others.filter((other): other is [element: FloatArray, transform?: Matrix] => other[0] !== null && other[0] !== undefined);
 
         if (nonNullOthers.length === 0) {
             return source;
         }
 
         if (!source) {
-            return this._mergeElement(nonNullOthers[0], nonNullOthers.slice(1));
+            return this._mergeElement(kind, nonNullOthers[0][0], nonNullOthers[0][1], nonNullOthers.slice(1));
         }
 
-        const len = nonNullOthers.reduce((sumLen, elements) => sumLen + elements.length, source.length);
+        const len = nonNullOthers.reduce((sumLen, elements) => sumLen + elements[0].length, source.length);
+
+        const transformRange = (element: FloatArray, matrix: Matrix | undefined, offset: number, length: number) => {
+            if (matrix) {
+                if (kind === VertexBuffer.PositionKind) {
+                    VertexData._TransformVector3Coordinates(element, matrix, offset, length);
+                } else if (kind === VertexBuffer.NormalKind) {
+                    VertexData._TransformVector3Normals(element, matrix, offset, length);
+                } else if (kind === VertexBuffer.TangentKind) {
+                    VertexData._TransformVector4Normals(element, matrix, offset, length);
+                }
+            }
+        };

```

Comment:
nit: not a big deal, but you can do the if checks outside instead of inside the function to improve perf slightly.

Reply:
```suggestion
        const transformRange = matrix ? () => {} :
            kind === VertexBuffer.PositionKind ? VertexData._TransformVector3Coordinates :
            kind === VertexBuffer.NormalKind ? VertexData._TransformVector3Normals :
            kind === VertexBuffer.TangentKind ? VertexData._TransformVector4Normals;
```